knative-operator/pkg/common/service_monitor.go

```diff
@@ -13,96 +12,102 @@ import (
-	SourceNameLabel                  = "eventing.knative.dev/sourceName"
-	SourceRoleLabel                  = "sources.knative.dev/role"
-	TestMonitor                      = "TEST_MONITOR"
-	TestSourceServiceMonitorPath     = "TEST_SOURCE_SERVICE_MONITOR_PATH"
-	TestSourceServicePath            = "TEST_SOURCE_SERVICE_PATH"
+	SourceLabel     = "eventing.knative.dev/source"
+	SourceNameLabel = "eventing.knative.dev/sourceName"
+	SourceRoleLabel = "sources.knative.dev/role"
 )
 
-func SetupSourceServiceMonitor(client client.Client, instance *appsv1.Deployment) error {
+func SetupSourceServiceMonitorResources(client client.Client, instance *appsv1.Deployment) error {
 	labels := instance.Spec.Selector.MatchLabels
-
 	clientOptions := mf.UseClient(mfclient.NewClient(client))
-	// create service for the deployment
-	manifest, err := mf.NewManifest(getMonitorPath(TestSourceServicePath, EventingSourcePath), clientOptions)
+	// Create service monitor service for source
+	smsManifest, err := createServiceMonitorServiceManifest(labels, instance.Name, instance.Namespace, clientOptions)
 	if err != nil {
-		return fmt.Errorf("unable to parse source service manifest: %w", err)
+		return err
 	}
-	transforms := []mf.Transformer{updateService(labels, instance.Name), mf.InjectOwner(instance), mf.InjectNamespace(instance.Namespace)}
-	if manifest, err = manifest.Transform(transforms...); err != nil {
+
+	if *smsManifest, err = smsManifest.Transform(mf.InjectOwner(instance)); err != nil {
 		return fmt.Errorf("unable to transform source service manifest: %w", err)
 	}
-	if err := manifest.Apply(); err != nil {
+	if err := smsManifest.Apply(); err != nil {
 		return err
 	}
-
-	// get service back, needed for the UID and setting owner refs
+	// Get service back, needed for the UID and setting owner refs
 	srv := &v1.Service{}
 	if err := client.Get(context.TODO(), types.NamespacedName{Name: instance.Name, Namespace: instance.Namespace}, srv); err != nil {
 		return err
 	}
-	// create service monitor for source
-	manifest, err = mf.NewManifest(getMonitorPath(TestSourceServiceMonitorPath, EventingSourceServiceMonitorPath), clientOptions)
+	// Create service monitor for source
+	smManifest, err := createServiceMonitorManifest(labels, instance.Name, instance.Namespace, clientOptions)
 	if err != nil {
-		return fmt.Errorf("unable to parse source service monitor manifest: %w", err)
-	}
-	transforms = []mf.Transformer{updateServiceMonitor(labels, instance.Name), mf.InjectOwner(srv), mf.InjectNamespace(instance.Namespace)}
-	if manifest, err = manifest.Transform(transforms...); err != nil {
-		return fmt.Errorf("unable to transform source service monitor manifest: %w", err)
+		return err
 	}
-	return manifest.Apply()
-}
 
-func getMonitorPath(envVar string, defaultVal string) string {
-	path := os.Getenv(envVar)
-	if path == "" {
-		return defaultVal
+	if *smManifest, err = smManifest.Transform(mf.InjectOwner(srv)); err != nil {
```

Comment:
Why not use the instance as an owner here too? Both are created from `instance` technically. The ServiceMonitor isn't owned by the `service`.

That way you could prevent the interim fetching and also create just one manifest as with the RBAC stuff.

Reply:
One way to do this is to have the service depend on the instance deployment and service monitor on the service because when the deployment is gone there is nothing to scrape from thus service is gone and then service monitor which depends on the existence of the service is gone. This is [how](https://github.com/operator-framework/operator-sdk/blob/53b00d125fb12515cd74fb169149913b401c8995/pkg/metrics/service-monitor.go#L72) it was done at the operator sdk side.

Reply:
Meh, I guess it's a detail anyway but I personally don't see a reason to build a requirement chain here as both resources are owned by the same thing, technically.

Reply:
Ok I can remove it no problem for me.